### PR TITLE
Validate and Persist SDK versions during compilation for access at run time

### PR DESCRIPTION
**Pull request by harvsu**
This is a new project to create a gradle plugin that validates the Mapbox SDK version to be RFC compliant(semver.org) and persists it in assets folder.

This is required because the Telemetry SDK need versions of all Mapbox library modules at runtime to generate the reformed user agent. All SDK versions now become part of the user agent for events sent to the backend.
 https://github.com/mapbox/mobile-telemetry/issues/452

This project is implemented in Groovy because Groovy is the official DSL for gradle. It is not much different from Java as it is also a JVM based language and Java-syntax-compatible object-oriented programming language.

This project also has the circleCI and Jacoco set up and tested on the current branch but will be made sure of the same on master branch in future pull requests.

This plugin should be applied across all Mapbox android SDK modules. Since all android teams that sends events through Telemetry SDK will be including this plugin in their SDK modules,  @mapbox/mobile-telemetry wants at least one review each from @mapbox/maps-android and @mapbox/navigation-android

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+          command: make checkstyle
+      - run:
+          name: Check codenarc code style
+          command: make codenarc
+      - run:
+          name: Run unit-test
+          command: make test
+      - run:
+          name: Build test plugin
+          command: ./gradlew assemble
```
Comment (tobrun) [inline .circleci/config.yml]: matches `make release` command

File: cloudformation/ci.template.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+var cf = require('@mapbox/cloudfriend');
```
Comment (tobrun) [inline cloudformation/ci.template.js:1]: the permissions in this cloudformation template don't match the permissions needed by this PR. eg. devicefarm

File: gradle/dependencies.gradle [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    dependenciesList = [
+            // unit test
+            junit: "junit:junit:${version.junit}",
+
+    ]
+
+    pluginDependencies = [
+            checkstyle: "com.puppycrawl.tools:checkstyle:${pluginVersion.checkstyle}",
+    ]
+}
```
Comment (tobrun) [inline gradle/dependencies.gradle:21]: formatting of this file doesn't match other .gradle files

File: gradle/mvn-push-plugin.gradle [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+def getRepositoryUsername() {
+    return hasProperty('NEXUS_USERNAME') ? NEXUS_USERNAME : ""
+}
+
+def getRepositoryPassword() {
+    return hasProperty('NEXUS_PASSWORD') ? NEXUS_PASSWORD : ""
+}
+
+
```
Comment (tobrun) [inline gradle/mvn-push-plugin.gradle]: nit additional line

File: gradle/mvn-push-plugin.gradle [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        from sources
+    }
+
+    task groovyDocs(type: Groovydoc) {
+        def sources = project.sourceSets.main.allGroovy
+        source = sources
+        classpath = project.sourceSets.main.output + project.sourceSets.main.compileClasspath
+    }
+
+
```
Comment (tobrun) [inline gradle/mvn-push-plugin.gradle]: nit additional line

File: gradle/mvn-push-plugin.gradle [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    artifacts {
+
+        archives generateSourcesJar
+        archives javaDocsJar
+        archives groovyDocsJar
+    }
+
+
```
Comment (tobrun) [inline gradle/mvn-push-plugin.gradle]: nit additional line

File: gradle/mvn-push-plugin.gradle [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    task javaDocsJar(type: Jar, dependsOn: javaDocs) {
+        classifier = 'javadoc'
+        from javaDocs.destinationDir
+    }
+
+
+    artifacts {
+
```
Comment (tobrun) [inline gradle/mvn-push-plugin.gradle]: nit additional line

File: gradle/mvn-push-plugin.gradle [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        source = sources
+        classpath = project.sourceSets.main.output + project.sourceSets.main.compileClasspath
+    }
+
+    task javaDocsJar(type: Jar, dependsOn: javaDocs) {
+        classifier = 'javadoc'
+        from javaDocs.destinationDir
+    }
+
+
```
Comment (tobrun) [inline gradle/mvn-push-plugin.gradle]: nit additional line

File: src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        if (outputFile.exists()) {
+            outputFile.delete()
+        }
+
+        outputFile.createNewFile()
+
+    }
+
+
+}
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy:68]: formatting of this file is inconsistent

File: src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Task to create a file with specified name and write specified SDK version into it.
+ */
+class SaveSDKVersionInfoTask extends DefaultTask {
+
+    /**
+     * Output directory set to this task.
+     */
+    @OutputDirectory
+    public File outputDir
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy]: public?

File: src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * Output file set to this task.
+     */
+    @OutputFile
+    public File outputFile
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy]: public?

File: src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+
+    /**
+     * SDK version set to this task.
+     */
+    public String sdkVersion
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy]: public?

File: src/main/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTask.groovy [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    @TaskAction
+    void action() {
+
+        if (outputDir != null && outputDir.exists()) {
+
+            outputDir.deleteDir()
+
+        }
+    }
+}
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTask.groovy:25]: formatting of this file?

File: src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+     * <p>
+     * @param version version string
+     * @return true if version conforms to regex pattern {@value #SEM_VER_REGEX}
+     */
+    @VisibleForTesting
+    static boolean validateVersion(String version) {
+
+        return Pattern.matches(SEM_VER_REGEX, version as CharSequence)
+    }
+}
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy:82]: formatting of this file?

File: src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * This plugin class persists the SDK version into a file named after the applicationId in assets folder.
+ * For every library variant in the project, this plugin reacts to resources generation
+ * part of build process and executes {@link SaveSDKVersionInfoTask}
+ *
+ */
+class PersistSDKVersionInfo implements Plugin<Project> {
+
+    public static final String LOG_TAG = "PersistSDKVersionInfo"
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy]: unused and public?

File: src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    public static final String SDK_VERSIONS_DIR = "/sdk_versions"
```
Comment (tobrun) [inline src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy]: public

File: src/test/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTaskTest.groovy [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        FileReader reader = new FileReader(saveSDKVersionInfoTask.outputFile)
+        String line = reader.readLine()
+        reader.close()
+
+        assertTrue(line == SDK_VERSION)
+
+    }
+
+
+}
```
Comment (tobrun) [inline src/test/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTaskTest.groovy:72]: formatting of this file is inconsistent

File: src/test/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTaskTest.groovy [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+        assertTrue(sdkVersionCleanUpTask.outputDir.exists())
+        assertTrue(sdkVersionCleanUpTask.outputDir.isDirectory())
+
+        sdkVersionCleanUpTask.action()
+
+        assertFalse(sdkVersionCleanUpTask.outputDir.exists())
+
+    }
+
```
Comment (tobrun) [inline src/test/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTaskTest.groovy:42]: formatting of this file is inconsistent

File: src/test/groovy/com/mapbox/android/sdk/versions/PersistVersionInfoTest.groovy [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+        assertFalse(PersistSDKVersionInfo.validateVersion("1.2.3.DEV"))
+        assertFalse(PersistSDKVersionInfo.validateVersion("1.2-SNAPSHOT"))
+        assertFalse(PersistSDKVersionInfo.validateVersion("1.2.31.2.3----RC-SNAPSHOT.12.09.1--..12+788"))
+        assertFalse(PersistSDKVersionInfo.validateVersion("1.2-RC-SNAPSHOT"))
+        assertFalse(PersistSDKVersionInfo.validateVersion("-1.0.3-gamma+b7718"))
+        assertFalse(PersistSDKVersionInfo.validateVersion("9.8.7+meta+meta"))
+        assertFalse(PersistSDKVersionInfo.validateVersion("9.8.7-whatever+meta+meta"))
+        assertFalse(PersistSDKVersionInfo.validateVersion("SNAPSHOT.12.09.1--------------------------------..12"))
+    }
+}
```
Comment (tobrun) [inline src/test/groovy/com/mapbox/android/sdk/versions/PersistVersionInfoTest.groovy:86]: formatting of this file is inconsistent

**Comment by tobrun** [comment]
> This project is implemented in Groovy because Groovy is the official DSL for gradle. It is not much different from Java as it is also a JVM based language and Java-syntax-compatible object-oriented programming language.

note that both internally at Mapbox as the community are moving towards adopting kotlin for this.

File: .circleci/config.yml (hunk 1 above)
Comment (harvsu) [reply .circleci/config.yml]: Good Catch! Will change it!

File: cloudformation/ci.template.js (hunk 2 above)
Comment (harvsu) [reply cloudformation/ci.template.js:1]: Got it. Removed devicefarm .

File: gradle/mvn-push-plugin.gradle (hunk 4 above)
Comment (harvsu) [reply gradle/mvn-push-plugin.gradle]: Removed additional line. Done.

File: gradle/mvn-push-plugin.gradle (hunk 5 above)
Comment (harvsu) [reply gradle/mvn-push-plugin.gradle]: Removed additional line. Done.

File: gradle/mvn-push-plugin.gradle (hunk 6 above)
Comment (harvsu) [reply gradle/mvn-push-plugin.gradle]: Removed additional line. Done.

File: gradle/mvn-push-plugin.gradle (hunk 7 above)
Comment (harvsu) [reply gradle/mvn-push-plugin.gradle]: Removed additional line. Done.

File: gradle/mvn-push-plugin.gradle (hunk 8 above)
Comment (harvsu) [reply gradle/mvn-push-plugin.gradle]: Removed additional line. Done.

File: src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy (hunk 10 above)
Comment (harvsu) [reply src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy]: Good find. Removed public access.

File: src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy (hunk 11 above)
Comment (harvsu) [reply src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy]: Removed public access.

File: src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy (hunk 12 above)
Comment (harvsu) [reply src/main/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTask.groovy]: Removed public access.

File: src/main/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTask.groovy (hunk 13 above)
Comment (harvsu) [reply src/main/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTask.groovy:25]: Fixed formatting. Done.

File: src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy (hunk 14 above)
Comment (harvsu) [reply src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy:82]: Fixed formatting. Done.

File: src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy (hunk 15 above)
Comment (harvsu) [reply src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy]: Removed LOG_TAG. Done.

File: src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy (hunk 16 above)
Comment (harvsu) [reply src/main/groovy/com/mapbox/android/sdk/versions/PersistSDKVersionInfo.groovy]: Changed it to private. Done.

File: src/test/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTaskTest.groovy (hunk 17 above)
Comment (harvsu) [reply src/test/groovy/com/mapbox/android/sdk/versions/SaveSDKVersionInfoTaskTest.groovy:72]: Fixed formatting. Done.

File: src/test/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTaskTest.groovy (hunk 18 above)
Comment (harvsu) [reply src/test/groovy/com/mapbox/android/sdk/versions/SDKVersionCleanUpTaskTest.groovy:42]: Fixed formatting. Done.

File: src/test/groovy/com/mapbox/android/sdk/versions/PersistVersionInfoTest.groovy (hunk 19 above)
Comment (harvsu) [reply src/test/groovy/com/mapbox/android/sdk/versions/PersistVersionInfoTest.groovy:86]: Fixed Formatting. Done.

**Review by alfwatt — APPROVED** [review]
Good job mopping up the formatting, approved!

**Pull request merged**